### ISSUE TITLE
async/await style library for simplified management of events

### ISSUE THREAD

**Issue by trentgill** [monome/crow#493]
one thing that i continually find frustrating with crow (and norns) is having to do everything in an event callback when that calls doesn't return instantly.

lua obviously doesn't support async/await natively, but we can get most of the way there with some metatables and coroutines. this is effectively what the `clock` library already does. so i propose some kind of library which allows for these kinds of actions without the hassle of managing complicated event handlers, especially when using functionality that has a shared callback for multiple queries (eg. `ii`).

the first thing is that we create a new function called `async` which is essentially an alias for `clock.run` with some added features. in a way this also helps clarify what's happening in `clock.run` as it essentially spawns a new thread which runs in parallel to the main thread.

so we can do the same thing you'd do in a clock routine:
```lua
async(function()
    clock.sleep(1)
    output[1].volts = math.random(0,5)
  end)
```

## extending clock.run() with looping constructs

very typically a `clock.run` block will wrap the contents with a `while true do` or some other loop construct that allows the block to run multiple times. this gets into similar territory to the `timeline` library, but perhaps that library could be superseded by this new `async` library -- or `timeline` could be updated to be all about lists of time-action pairs, while the `async` lib takes care of the back-end sequencing.

```lua
async.loop(function(i) -- note we give access to the loop iterator
    clock.sleep(1)
    output[1].volts = i % 5
  end)


async.times(4, function(i) -- run the async function a fixed number of times (for i=1,4 do)
    clock.sleep(1)
    output[1].volts = i % 5
  end)

local out_cycle = {1 = 0, 2 = 2, 3 = -3, 4 = -1}
async.iter(pairs(out_cycle), function(k,v) -- as 'times' but accept a k-v generating iterator (for k,v in pairs(mytab) do)
    clock.sleep(1)
    output[k].volts = v
  end)

async.pred(math.random, function() -- continues to loop until the predicate function returns nil or zero
    clock.sleep(1)
    output[1].volts = math.random(-3,3)
  end)
```

## adding await() to allow asynchronous calls without an event handler

this is all generally useful, but really just syntactic sugar over the clock library. it only gets interesting when we introduce `await` functions which pause execution until they receive a value. the primary use case on crow is for `ii` queries, but we could use it for anything that returns values over time (eg. `input` table):

```lua
async.loop(function()
    output[1].volts = await(ii.jf, 'ramp')
  end)
```

the big win here is that we never have to define an event handler for `ii.jf.event` and all the matching of query to action is handled implicitly. the downside is perhaps a convoluted setup to get this to happen (though not more complicated than a clock routine).

even though it might feel a little complex, the idea is that if we can make all of the event handlers function in this way, then there's only a single set of syntax that one needs to learn to be able to access all of this kind of functionality. it could also be used to handle callbacks more generally, ala crow as a follower (currently ii.self.x):

```lua
-- query txi's 2nd knob and apply it's response to output[2]
async.loop(function()
    output[2].volts = await(ii.txi, param, 2)
  end)

-- handle all received messages from another crow or teletype: eg. ii.crow.call2(1,4)
async.loop(function()
    c1, c2 = await(ii.self.call2)
    output[c1].volts = c2
  end)
```

this requires the `await` function to be kind of fancy under-the-hood as it would need to be modified to support all `ii` devices (we can code-gen this like we do the rest of the lib), and use a system to match queries to their responses. a simple generated ID number would allow us to thread the call and response together.

---

## asl, in the abstract

presently ASL is restricted to working on outputs, and only 1 per output. they can't be nested and they can't interact. i would love to see this overhauled such that ASL is a more general idea of creating continuous functions in time, such that they can be applied to anything. the initial use-case for this is to replicate a function i wrote a long time ago called `fnl()` which was for writing ASL-style descriptions of modulations that get sent to an `ii` destination.

the idea here is to allow ASL to stand on its own, and pass that object to different 'players' that can run it. the obvious case is for a CV output, which could use similar or the same syntax as currently. but we also want to be able to define an ASL and have it live in a variable:

```lua
myAsl = loop{to(5,1,'exp'), to(0,10,'log')}
```

then we have some mechanism to run the ASL in a step-wise fashion. presently we just pass it to the C-layer which does all the magic for us at the fixed sample-rate. by allowing the sample-rate to be defined in the lua layer, we can use those same descriptions but call any lua function we want that uses it:

```lua
-- define our simple A-R envelope ASL
myAsl = loop{to(5,1,'exp'), to(0,10,'log')}

-- create an "async player" that takes an ASL, sets the sample rate to 100ms, and sends the value to an ii destination every tick.
async.asl(myAsl, async.sleep(0.1), ii.wsyn.curve)
```

i'm not confident of the syntax here, as i think there could be something more elegant that just having a "timing" parameter. perhaps some kind of meta-table system that allows for a timing object to be connected to a function call:

```lua
throttle(0.1) >> ii.wsyn.curve
```

this is very much uncertain to me right now and requires some deeper thought about how the rest of the system would function, but i wanted to write it down here as a jumping off point. i think it largely depends on how we plan to handle other data-generator functions (like `input`).

## sequins everywhere

i won't write examples, but it would be wild and not-too-hard to make all of the above elements be sequins enabled. the timing parameter could use an overloaded `sync` function that looks something like:

```lua
sync{2/3, 1/3} -- performs clock.sync{2/3} then clock.sync{1/3} in an alternating fashion
```

## input handling & analysis objects

presently we use a 2-step process to access the different input modalities. while there is some backend utility to switching between different processing, we could infer it from a more general syntax.

my initial thought is that if the input object could have an "analysis" object attached, we could simplify the behaviour:

```lua
-- apply the quantized input[1] to output[1] whenever the note changes
async.loop(function()
    output[1].volts = input[1] >> scale{0,2,4,7,9}
    output[2](ar()) -- when the value changes, we trigger an AR envelope on output[2]
  end)

-- every 100ms query the second txi knob, and only output when the value has changed to a new window
async.loop(function()
    clock.sleep(0.1)
    mySelection = await(ii.txi, 'param', 2) >> window{0,1,2,3,4,5}
  end)

-- the above may not be possible, so we could add a timing object in sequence
async.loop(function()
    -- method style
    mySelection = await(ii.txi, 'param', 2):throttle(0.1) >> window{0,1,2,3,4,5}
    -- overloaded style
    mySelection = await(ii.txi, 'param', 2) >> throttle(0.1) >> window{0,1,2,3,4,5}
  end)
```

it gets kinda weird and funky, and i'm not sure overloading the bit-shift-right operator is the best idea, but it allows us to configure this kind of thing while sticking with a pure lua implementation.

the value of abstracting the "analysis" objects from the underlying hardware is that we could re-use identical syntax (even the same object) to eg quantize any variable in the system. specifically imagine generating a random data in a clock routine & being able to quantize it exactly the same way as input[n].scale or output[n].scale work, but send it to an `ii` destination instead.

## wrap up

there's much more to be explored here, and i haven't even begun to think about how it could apply to norns. my hope is that it provokes some conversation about an alternative system for managing events & callbacks, and leads toward some kind of new syntax that feels more unified than the current system.

presently i find there is a lot of mental overhead remembering how each family of functions works and which things are declarative vs imperative. the above is just one of many approaches that i've been thinking about, and it feels like the smallest jump from the current style. i'm particularly interested in unifying the syntax so that everything looks like a clock routine -- not because clock routines are the most elegant form (i *wish* there was a shorthand for writing inline functions), but because they feel the most flexible & universal.

lastly, i'm not suggesting we totally replace the existing syntax and break all the existing scripts - that would be ridiculous. but instead dreaming up a new style that would allow new users a greatly streamlined learning process for interacting with these sound machines. i'm using the crow backend on numerous new projects and the idea of asking musicians to learn the whole stack of functionality as well as syntax is a huge ask -- it restricts the proposition to those that really want to get their hands dirty. what if there was just one style - a single way of describing all the core things you want to do with your music, and you don't have to learn a whole (yet tiny) programming language to do it.

it's pie in the sky, sure, but i do think there's a way to make it feel more immediate (tidal cycles comes to mind), while allowing access to a complete lua implementation for those that want to do more than plug a few functions together.

happy sunday :)
